### PR TITLE
GPU: Added code to unswizzle textures and ported the surface viewer from citra

### PR DESCRIPTION
**Pull request by Subv**
Things left to do:

- [ ] Detect when textures are not swizzled (There seems to be a register for this)
- [ ] Decode the textures to a format supported by the host GPU
- [ ] Analyze the shaders to know which textures are actually used

### CODE REVIEW

File: src/video_core/debug_utils/debug_utils.cpp [hunk 1]
```diff
@@ -0,0 +1,67 @@
...
+// Refer to the license.txt file included.
+
+#include <algorithm>
+#include <condition_variable>
+#include <cstdint>
+#include <cstring>
+#include <fstream>
+#include <map>
+#include <mutex>
+#include <stdexcept>
```
Comment (lioncash) [inline src/video_core/debug_utils/debug_utils.cpp]: As far as I can tell, nothing uses stdexcept

File: src/video_core/debug_utils/debug_utils.cpp [hunk 2]
```diff
@@ -0,0 +1,67 @@
...
+#include "common/common_types.h"
+#include "common/file_util.h"
+#include "common/logging/log.h"
+#include "common/math_util.h"
+#include "common/vector_math.h"
+#include "video_core/debug_utils/debug_utils.h"
+
+namespace Tegra {
+
+std::shared_ptr<DebugContext> g_debug_context;
```
Comment (lioncash) [inline src/video_core/debug_utils/debug_utils.cpp]: Can we avoid introducing globals?

Comment (Subv) [reply src/video_core/debug_utils/debug_utils.cpp]: This code was ported almost verbatim from citra.
The reason this is a global and not a member of GPU is that a reference to it is stored within each BreakpointObserver subclass, and these are constructed at startup before the GPU instance is even created.

Comment (lioncash) [reply src/video_core/debug_utils/debug_utils.cpp]: Is that truly the best way to do it though? Given this is introducing it, should we not try and do this in a better way that doesn't specifically rely on it being global before we build anything on top of it?

Like, yeah, it's from citra, but that doesn't mean it's necessarily a good idea to perform a verbatim copy. This just introduces the same mistakes made previously.

**Comment by Hexagon12** [comment]
@Subv Can you please rebase this PR, since it has conflicts?

**Pull request merged**